File: docs/algorithms/code/algorithms/maxDepth1.ts
function maxDepth(root: TreeNode | null): number {
  if (!root) {
    return 0;
  }

  return Math.max(maxDepth(root.left), maxDepth(root.right)) + 1;
}

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

interface TreeNodeItem {
  node: TreeNode;
  depth: number;
}

function maxDepth(root: TreeNode | null): number {
  if (!root) {
    return 0;
  }

  const stack: Array<TreeNodeItem> = [];
  stack.push({
    node: root,
    depth: 1,
  });

  let max = 1;

  while (stack.length) {
    const treeNodeItem = stack.shift();
    max = Math.max(treeNodeItem.depth, max);

    if (treeNodeItem.node.left) {
      stack.push({
        node: treeNodeItem.node.left,
        depth: treeNodeItem.depth + 1,
      });
    }

    if (treeNodeItem.node.right) {
      stack.push({
        node: treeNodeItem.node.right,
        depth: treeNodeItem.depth + 1,
      });
    }
  }

  return max;
}

File: docs/algorithms/code/algorithms/maxDepth1.go
// 思路：最大深度就是左右子树的最大深度加1
// 复杂度分析：时间复杂度 O(n)，空间复杂度 O(h)
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}

	left := maxDepth(root.Left)
	right := maxDepth(root.Right)
	if left > right {
		return left + 1
	}
	return right + 1
}